core/shared/src/main/scala/scalaz/zio/Fiber.scala

```diff
@@ -74,11 +70,11 @@ trait Fiber[+E, +A] { self =>
   final def zipWith[E1 >: E, B, C](that: => Fiber[E1, B])(f: (A, B) => C): Fiber[E1, C] =
     new Fiber[E1, C] {
       def observe: IO[Nothing, ExitResult[E1, C]] =
-        self.observe.seqWith(that.observe)(_.zipWith(_)(f))
+        self.observe.seqWith(that.observe)(_.zipWith(_)(f, _ ++ _))
```

Comment:
ha, again my mistake 😅

Reply:
My mistake for sloppy types. 😆 I think it would be great if we modified `++` so that you can only combine `l` and `r` when either `l: Cause[Nothing]` or `r: Cause[Nothing]`. 

The reason being: a fiber cannot fail for 2 different `E`. 

It's possible to solve this with some effort:

```scala
trait OneFailure[E1, E2] {
      type Output

      def left: E1 => Output
      def right: E2 => Output

      def proof: Either[E1 => Nothing, E2 => Nothing]
    }
    trait OneFailureInstances0 extends OneFailureInstances1 {
      implicit def leftFailure[E]: Aux[E, Nothing, E] =
        new OneFailure[E, Nothing] {
          type Output = E

          def left: E => Output = identity
          def right: Nothing => Output = identity

          def proof = Right(identity)
        }
    }
    trait OneFailureInstances1 {
      type Aux[A, B, C] = OneFailure[A, B] {
        type Output = C
      }
      implicit def rightFailure[E]: Aux[Nothing, E, E] =
        new OneFailure[Nothing, E] {
          type Output = E

          def left: Nothing => Output = identity
          def right: E => Output = identity

          def proof = Left(identity)
        }
    }
    object OneFailure extends OneFailureInstances0 {
      def concat[A, B, C](l: List[A], r: List[B])(implicit ev: OneFailure.Aux[A, B, C]): List[C] =
        ???

      concat(List.empty[Nothing], List[Int](1, 2, 3)): List[Int]
      concat(List[Int](1, 2, 3), List.empty[Nothing]): List[Int]
      concat(List.empty[Nothing], List.empty[Nothing]): List[Int]
    }
```

Then you will get a type error if you try to `++` two `Cause` but one of them is not known to have `Nothing` for its `E`.

This would have prevented the above implementation, but would still allow use of `++` to concatenate finalizer errors onto a primary cause.

Reply:
Would there be any disadvantage in moving `++` right into `OneFailure`?
```
class OneFailure e1 e2 f | e1 e2 -> f where (++) :: e1 -> e2 -> f
instance OneFailure (Cause Nothing) (Cause Nothing) (Cause Nothing) ...
instance OneFailure (Cause Nothing) (Cause e) (Cause e) ...
instance OneFailure (Cause e) (Cause Nothing) (Cause e) ...
```
I know these overlap 😓